Add design goals and threat model (#55)
boltlabs-inc/key-mgmt-spec: design-philosophy-and-threat-model.md

# Design Philosophy and Threat Model
[Philosophy](#philosophy)<br>
[Threat Model](#threat-model-and-mitigations)<br>

## Philosophy
An informal and non-exhaustive list:

- [Zero Trust](https://csrc.nist.gov/publications/detail/sp/800-207/final).

>     - defense in depth, 
>     - graceful failures, 
>     - distribution of trust, no single point of failure, 
>     - least privilege, 

Reviewer: ```suggestion
    - defense in depth 
    - graceful failures 
    - distribution of trust, no single point of failure 
    - least privilege 
```

    - minimize attack surface
- Acknowledgment of human fallibility, importance of usability
- Trust-but-verify, open source, abort & identify malicious behavior
- Consideration of challenges to security posed by integration with the service provider’s asset management product: 

>     - Our software should be conducive to integration in a way that protects against threats not in our immediate scope; 

Reviewer: ```suggestion
    - Our software should be conducive to integration in a way that protects against threats not in our immediate scope 
```

    - The importance of providing clear, actionable guidance for organizational partners
- Extensibility, modularity 
    - Including allowing for transition to more robust threat models, e.g., fault-tolerant Byzantine servers
    - Generalizability
    - Keep components modular so that we can serve different types of community members and service providers with different constraints/end goals. (e.g., Do not preclude asset owner from running their own server; allow asset owners to choose from a menu of custodial options; allow service providers to choose from a menu of deployment options)
- Deployment flexibility:
    - Support on-premises HSMs and in-cloud servers

>     - Allow use of external cloud infrastructure while minimizing trust required by using secure processors and/or secure hardware.

Reviewer: ```suggestion
    - Allow use of external cloud infrastructure while minimizing trust required by using secure processors and/or secure hardware
```

For consistency

A core principle in the development of Lock Keeper is to provide guidance and make decisions that enable the consuming application to create a secure and usable product.

[External Actors](#external-actors)<br>
[Internal Actors](#internal-actors) <br>
[Software Dependencies](#software-dependencies)<br>
[Additional Threat Considerations](#additional-threat-considerations)

### External actors

#### Passive and active network-level adversaries
We advise consideration of the following:

1. Denial of service (DoS) attacks, including both temporary and permanent (e.g., volumetric, loss of server): For deployment,  standard DoS mitigations at both the cryptographic and the networking level are advisable, including mechanisms that account for server compromise and failure, namely

>     1. Server over-provisioning; 
>     1. Server replication; 
>     1. Automated volumetric DoS protection; and
>     1. Cold backups, where/if appropriate.

Reviewer: ```suggestion
    1. Server over-provisioning 
    1. Server replication 
    1. Automated volumetric DoS protection
    1. Cold backups, where/if appropriate
```

Author: I don't intend to leave off punctuation in lists in general. The "philosophy" section needs to be expanded as part of a separate ticket. Planning to write the ticket after merging this PR.

#### Other humans and malware

Malware and device compromise is a consideration of Lock Keeper, particularly with respect to confidentiality assurance:

1. Lock Keeper should provide confidentiality of the key and other sensitive user data on device and in transit:
    1. A (whole) secret should never be held by any single entity, except possibly by the asset owner after an explicit, authorized request. Certain types of keys may not have full recovery functionality enabled, e.g., if the asset is considered a shared asset with the service provider. Threshold cryptography/MPC solutions should be leveraged to this end for key generation, distribution, and for signing/decryption.
    1. The asset owner must authenticate in order to recover, use, or otherwise interact with their key. Care must be taken that the authentication mechanism is robust against impersonation attacks, e.g., by use of threshold cryptography/MPC solutions and/or a distributed PAKE.
1. The developers integrating Lock Keeper should receive integration guidance on how to avoid decisions that undermine key or user data confidentiality.

> Humans with whom the digital asset owner interacts physically, e.g., family members, housemates, friends, and coffee shop dwellers, or virtually. Specifically, the software integrating Lock Keeper should include mitigations for the following types of threats:

Reviewer: This looks like a bullet point but I think it's meant to be context for the following bullet points. Maybe rearrange the sentences something like this:

```suggestion
 The software integrating Lock Keeper should include mitigations against humans with whom the digital asset owner interacts physically, e.g., family members, housemates, friends, and coffee shop dwellers, or virtually. Specifically, it should consider the following types of threats:
```
